src/models/models.cpp

```diff
@@ -40,4 +28,141 @@ void Models::list(std::string project_token, std::map<std::string, std::string>
+    console::error("There was an error downloading your model \n");
+  }
+}
+
+std::string Models::tags_to_string(Json tags){
+  auto tags_list = tags.array_items();
+  std::string model_tags;
+  if (tags_list.size() == 0){
+    model_tags = "nil";
+  } else {
+    for (int j = 0; j < tags_list.size(); j++) {
+      if(j == tags_list.size() - 1){
+        model_tags = model_tags + " " + tags_list[j].string_value();
+      } else {
+        model_tags = model_tags + " " + tags_list[j].string_value() + ",";
+      }
+    }
+  }
+  return model_tags;
+}
+
+void Models::download_to_file(std::string project_token, std::string display_name, std::string version, std::string output_path){
+
+    std::string model_url = DOWNLOAD_URL + "/projects/" + project_token + "/models?name=" + display_name + "&version=" + version;
+
+    std::string output_file = version + "_" + display_name + ".txt";
+    std::string download_path;
+
+    download_path = resolve_download_path(output_path, output_file);
+
+    console::debug("Download path: " + download_path);
+
+    Request::download(model_url, download_path);
+    
+    try {
+      YAML::Node model_error = YAML::LoadFile(download_path);
+      if (model_error["Error"]){
+        console::error("There was an error downloading your model to " + download_path + ": " + model_error["Error"].as<std::string>());
+      } else{
+        console::error("There was an error downloading your model to " + download_path);
+      }
+      FileManager::delete_file(download_path);
+    } catch (YAML::ParserException& e){
+      console::info("Model successfully downloaded to " + download_path);
+    } catch (...){
+      console::error("There was an error downloading your model to " + download_path);
+    }
+}
+
+std::string Models::resolve_download_path(std::string output_path, std::string output_file) {
+  std::string download_path;
+
+  if (output_path.compare(".") == 0){
+      download_path = FileManager::cwd() + "/" + output_file;
+    } else {
+      if (FileManager::is_dir(output_path)){
+        string dir = output_path.substr(output_path.find_last_of("/"));
+        if (dir.compare("/") == 0){
+          download_path = output_path + output_file;
+        } else{
+          std::size_t found = output_path.find(".");
+          if (found!=std::string::npos){
+            string file_ext  = output_path.substr(output_path.find_last_of("."));
+            if (file_ext.compare(".") == 0){
+              download_path = output_path + "/" + output_file;
+            } else {
+              if (FileManager::file_exists(output_path) || file_ext.size() <= 4){
+                download_path = output_path;
+              } else {
+                download_path = output_path + "/" + output_file;
+              }
+            }
+          } else{
+            download_path = output_path + "/" + output_file;
+          }
+        }
+      } else {
+        download_path = output_path;
+      }
+    }
+    return download_path;
```

Comment:
This whole function is hard to follow. Nested if statements inside nested if statements generally are not easy to follow. If it's easy to break this out to be easier to follow, please do so, otherwise it can pass for now.